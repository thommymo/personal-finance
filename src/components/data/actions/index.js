export const REQUEST_EXCHANGE_RATES = 'REQUEST_EXCHANGE_RATES'
export const RECEIVE_EXCHANGE_RATES = 'RECEIVE_EXCHANGE_RATES'
export const SET_PORTFOLIO_SELECTION = 'SET_PORTFOLIO_SELECTION'

export function requestExchangeRates(toCurrency) {
  return {
    type: REQUEST_EXCHANGE_RATES,
    toCurrency
  }
}

export function receiveExchangeRates(toCurrency, json) {
  return {
    type: RECEIVE_EXCHANGE_RATES,
    toCurrency,
    rates: json,
    receivedAt: Date.now()
  }
}

export function setPortfolioSelection(holdingsType,color) {
  return {
    type: SET_PORTFOLIO_SELECTION,
    holdingsType,
    color
  }
}

//TODO: ERROR HANDLING

export function fetchExchangeRates(toCurrency) {
  return function (dispatch) {
    dispatch(requestExchangeRates(toCurrency))
    return fetch(`https://api.fixer.io/latest?base=${toCurrency}`)
      .then(
        response => response.json(),
        error => console.log('An error occurred.', error))
        .then(json => {
          let currencyRates = json.rates
          currencyRates[toCurrency] = 1
          return dispatch(receiveExchangeRates(toCurrency, currencyRates))
        })
  }
}
