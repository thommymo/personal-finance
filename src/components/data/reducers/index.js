import { combineReducers } from 'redux'
import {
  REQUEST_EXCHANGE_RATES,
  RECEIVE_EXCHANGE_RATES,
} from '../actions'

function exchangeRates(
  state = {
    isFetching: false,
    items: []
  }, action
) {
  switch(action.type){
    case REQUEST_EXCHANGE_RATES:
      //Load the Values from API
      return {
        ...state,
        isFetching: true
      }
    case RECEIVE_EXCHANGE_RATES:
      //Load the Values from API
      console.log(action)
      return {
        ...state,
        rates: action.rates,
        isFetching: false
      }
    default:
      return state
  }
}

const rootReducer = combineReducers({
  exchangeRates,
})

export default rootReducer
