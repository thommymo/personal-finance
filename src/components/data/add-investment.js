import React, { Component } from 'react'
import {connect} from 'react-redux'
import styled from 'styled-components'
import { addHolding } from '../../actions'
import { holdingsWithMarketPrice } from '../../data/data'

/*
  TODO: Data Validation in Form
*/

class AddInvestment extends Component {
  constructor(props){
    super(props)
    this.state = {
      type: "Cash",
      name: "",
      y: 0,
      currency: "CHF",
      interest: 0,
      exchange: "NYSE",
      symbol: "",
      show: false
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
    this.handleClick = this.handleClick.bind(this)
    this.onEscape = this.onEscape.bind(this)
  }

  onEscape({ keyCode }) {
    if (keyCode === 27) {
      this.setState({
        type: "Cash",
        name: "",
        y: 0,
        currency: "CHF",
        interest: 0,
        exchange: "NYSE",
        symbol: "",
        show: false
      })
    }
  }

  handleChange(event){
    const name = event.target.name
    const value = (name === "y" || name === "interest") ? parseFloat(event.target.value) : event.target.value
    this.setState({
      [name]: value
    })
  }

  handleSubmit(){
    this.setState({
      show:false
    })
    this.props.dispatch(addHolding(this.state))
    this.setState({
      type: "Cash",
      name: "",
      y: 0,
      currency: "CHF",
      interest: 0,
      exchange: "NYSE",
      symbol: "",
    })
  }

  handleClick(){
    this.setState({
      show:true
    })
  }

  componentDidMount() {
    document.addEventListener('keydown', this.onEscape);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.onEscape);
  }

  render() {
    const currencies = Object.keys(this.props.exchangeRates.rates)
    return (
      <StyledAddInvestment>
        { !this.state.show &&
          <Button><button onClick={this.handleClick}>Add Investment</button></Button>
        }
        { this.state.show &&
          <Overlay>
            <FormWrapper>
              <h1>Add Investment</h1>
              <StyledForm onSubmit={this.handleSubmit}>
                <FormElementWrapper>
                  <label htmlFor="type">
                    Type
                  </label>
                  <select value={this.state.type} name="type" id="type" onChange={this.handleChange}>
                    <option value="Cash">Cash</option>
                    <option value="Equities">Equities</option>
                    <option value="Real Estate">Real Estate</option>
                    <option value="Coop">Coop</option>
                    <option value="Retirement Money">Retirement Money</option>
                    <option value="Private Loan">Private Loan</option>
                    <option value="Natural Resources">Natural Resources</option>
                    <option value="Other">Other</option>
                  </select>
                </FormElementWrapper>
                <FormElementWrapper>
                  <label htmlFor="name">
                    Name
                  </label>
                  <input type="text" name="name" value={this.state.name} id="name" onChange={this.handleChange} />
                </FormElementWrapper>
                <FormElementWrapper>
                  <label htmlFor="y">
                    { !holdingsWithMarketPrice.includes(this.state.type) &&
                `Value in ${this.state.currency}`
                    }
                    { holdingsWithMarketPrice.includes(this.state.type) &&
                `Number of Holdings`
                    }
                  </label>
                  <input type="text" name="y" value={this.state.y} id="y" onChange={this.handleChange} />
                </FormElementWrapper>
                <FormElementWrapper>
                  <label htmlFor="currency">
                    Currency
                  </label>
                  <select value={this.state.currency} name="currency" id="currency" onChange={this.handleChange}>
                    { currencies.map((currency) => (
                      <option key={currency} value={currency}>{currency}</option>
                    ))}
                  </select>
                </FormElementWrapper>
                { !holdingsWithMarketPrice.includes(this.state.type) &&
                  <FormElementWrapper>
                    <label htmlFor="interest">
                      Interest Rate
                    </label>
                    <input type="text" name="interest" value={this.state.interest} id="interest" onChange={this.handleChange} />
                  </FormElementWrapper>
                }
                { holdingsWithMarketPrice.includes(this.state.type) &&
                  <FormElementWrapper>
                    <label htmlFor="exchange">
                      Exchange
                    </label>
                    <select value={this.state.exchange} name="exchange" id="exchange" onChange={this.handleChange}>
                      <option value="NYSE">NYSE</option>
                      <option value="SWX">SWX</option>
                    </select>
                  </FormElementWrapper>
                }
                { holdingsWithMarketPrice.includes(this.state.type) &&
                  <FormElementWrapper>
                    <label htmlFor="symbol">
                      Symbol
                    </label>
                    <input type="text" name="symbol" value={this.state.symbol} id="symbol" onChange={this.handleChange} />
                  </FormElementWrapper>
                }
                <ButtonRight><input type="submit" value="Submit" /></ButtonRight>
              </StyledForm>
            </FormWrapper>
          </Overlay>
        }
      </StyledAddInvestment>
    )
  }

}

const mapStateToProps = (state, props) => (
  {...state}
)

export default connect(mapStateToProps)(AddInvestment)

const Overlay = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  background-color: rgba(255,255,255,0.98);
  height: 100vh;
  width: 100vw
`

const FormWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: 100vh;
`

const StyledForm = styled.form`
  display: flex;
  flex-direction: column;
`

const FormElementWrapper = styled.div`
  min-width: 300px;
  padding-bottom:30px;
  flex-direction: row;
  display: flex;
  align-items: center;
  justify-content: space-between;
`

const Button = styled.div`
  input, button {
    width:100%;
    background-color: white;
    color: black;
    border: 2px solid black;
    padding: 10px 24px;
    -webkit-transition-duration: 0.4s; /* Safari */
    transition-duration: 0.4s;
    align-self: center;
    font-size: 16px;
    :hover {
        background-color: black;
        color: white;
    }
  }
`
const ButtonRight = Button.extend`
  display: flex;
  justify-content: flex-end;
  margin:auto;
`
const StyledAddInvestment = styled.div`
  padding-bottom: 20px;
`
