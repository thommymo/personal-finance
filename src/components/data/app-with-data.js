import React, { Component } from 'react'
import { portfolio, holdingsWithMarketPrice } from '../../data/data'
import PieChart from '../molecules/piechart'
import { connect } from 'react-redux'
import TableWithHoldings from '../molecules/tablewithholdings'
import TableWithHoldingsFinMarkets from '../molecules/tablewithholdingsfinmarkets'
import { setPortfolioSelection } from '../../actions'
import LineChart from '../molecules/linechart'
import AddInvestment from '../data/add-investment'
import styled from 'styled-components'

//Data will come from a database later, instead of the import from "../../data/data"

/*
TODO
//Check which steps are really necessary
2. If no Data is available: Show error message
3. If no Data in IndexedDB, but data is loading: Show Loading message
4. If Data in IndexedDB, show with old data and load new data in background
5. If New Data from API, show data
*/

class AppWithData extends Component {

  setPortfolioSelection(holdingsType,color){
    this.props.dispatch(setPortfolioSelection(holdingsType,color))
  }

  render() {
    const { holdingsType, color } = this.props.selection
    const { marketDataForHoldings, portfolio } = this.props
    const shareValue = Object.keys(marketDataForHoldings.items).reduce((acc, symbol)=>{
      if(Object.keys(marketDataForHoldings.items[symbol]).length>0){
        const data = marketDataForHoldings.items[symbol]["Monthly Adjusted Time Series"]
        const allDates = Object.keys(data).sort((a,b)=>(a > b))
        acc[symbol] = parseFloat(data[allDates[(allDates.length-1)]]["5. adjusted close"])
      }
      return acc
    },{})

    const exchangeRates = this.props.exchangeRates.rates
    const filteredSymbols = portfolio.items.filter(holding => (holding.type === holdingsType)).map(holding => holding.symbol)
    const filteredMarketData = filteredSymbols.reduce((acc,symbol)=>{
        let marketDataForSymbol = this.props.marketDataForHoldings.items[symbol]
        if(marketDataForSymbol !== undefined){
          acc[symbol] = marketDataForSymbol
        }
        return acc
      },{})
    return (
      <ChartWrapper>
        { exchangeRates.isFetching &&
          //TODO: Add a beautiful Loading state
          <div>Loading</div>
        }
        { !exchangeRates.isFetching && !exchangeRates &&
          //TODO: Add a beautiful Error state
          <div>Error</div>
        }
        { !exchangeRates.isFetching && exchangeRates &&
          <PieChart
            filter={holdingsType}
            setPortfolioSelection={(holdingsType,color)=>this.setPortfolioSelection(holdingsType,color)}
            portfolio={portfolio.items}
            currency={exchangeRates}
            holdingsWithMarketPrice={holdingsWithMarketPrice}
            shareValue={shareValue}
          />
        }
        <ButtonWrapper><AddInvestment /></ButtonWrapper>
        { !exchangeRates.isFetching && exchangeRates && holdingsType && !holdingsWithMarketPrice.some(holding => holding===holdingsType) &&
          <TableWithHoldings
            portfolioSelection={holdingsType}
            color={color}
            portfolio={portfolio.items}
            currency={exchangeRates}
          />
        }
        { !exchangeRates.isFetching && exchangeRates && holdingsType && holdingsWithMarketPrice.some(holding => holding===holdingsType) &&
          <div>
            <TableWithHoldingsFinMarkets
              portfolioSelection={holdingsType}
              color={color}
              shareValue={shareValue}
              portfolio={portfolio.items}
              exchangeRates={exchangeRates}
            />
            <LineChart loadingStatus="loaded" data={filteredMarketData} color={color} holdingsType={holdingsType}/>
          </div>
        }

      </ChartWrapper>
    )
  }
}

const mapStateToProps = (state, props) => ({
  ...state
});

export default connect(mapStateToProps)(AppWithData)

const ChartWrapper = styled.div`
  display: block;
  flex-direction: column;
  align-items:center;
  width: 100vw;
`
const ButtonWrapper = styled.div`
  display: flex;
  width:100%;
  align-items:center;
  justify-content: center;
`
