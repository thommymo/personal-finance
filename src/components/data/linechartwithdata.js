import React, { Component } from 'react'
import idb from 'idb'
import LineChart from '../molecules/linechart'
import PieChartWithData from '../../components/data/piechart-with-data'
import '../../utils/global-css'
import { portfolio, holdingsWithMarketPrice } from '../../data/data'
import { connect } from 'react-redux'
import { fetchExchangeRates } from '../../components/data/actions'



/*
TODO:
1. New Structure (Data and Visual representation should be separated)
2. Clean Coding Principles
3. Everything should be tested
*/

class LineChartWithData extends Component {
  constructor(){
    super()
    this.state =Â {
      shareprice: "loading",
      symbol: "JNK",
      allData: {},
      symbols: []
    }
  }

  componentDidMount(){
    var portfolio=this.props.symbols
    console.log(portfolio);

    //foreach holding with market data, get Data and put it into client side storage
    var symbols = portfolio.filter(holding => holdingsWithMarketPrice.some(holdingSymbol => holdingSymbol === holding.type))

    this.getData(symbols).then(data => {
      for(let share of Object.keys(data)){
        data[share].then(data => {
          let newData = this.state.allData
          newData[share] = data
          let symbols = this.state.symbols
          symbols.push(share)
          this.setState({shareprice: "loaded", allData: newData, symbols: symbols})
          }
        )
      }
      //TODO: Exception handling
    })
  }

  //Get Data from IndexedDB
  //Let's assume for now that the data is never updated (TODO: Do this when data updates)

  getData = (symbols) => {
    const convertSWXData = (json) => {
      var object = {
        "Meta Data": {
          "1. Information": "Monthly Adjusted Prices and Volumes",
          "2. Symbol": json.valors[0].ISIN, //json.valors.ISIN
          "3. Last Refreshed": Date.now(), //delayedDateTime
          "4. Time Zone": "CET/Zurich" //CH/Zurich
        },
        "Monthly Adjusted Time Series": {
          // "2000-02-29": {  //valors.date.Date (20040924)
            // "1. open": "104.0000", //valors.date.Open
            // "2. high": "119.9400",//valors.date.High
            // "3. low": "97.0000", //valors.date.Low
            // "4. close": "114.6200",  //valors.date.Close
            // "5. adjusted close": "3.6693", //valors.date.Close
            // "6. volume": "65355200", //valors.date.TotalVolume
            // "7. dividend amount": "0.0000"
        }
      }

      let value = {}
      const dates = json.valors[0].data.Date
      for(let i=0; i<dates.length; i++){
        let date = String(dates[i])
        let nextdate = String(dates[(i+1)])
        if(date.substring(4,6)!==nextdate.substring(4,6)){
          date = date.substring(0,4) + "-" + date.substring(4,6) + "-" + date.substring(6,8)
          value[date] = {
            "1. open": json.valors[0].data.Open[i], //valors.date.Open
            "2. high": json.valors[0].data.High[i],//valors.date.High
            "3. low": json.valors[0].data.Low[i], //valors.date.Low
            "4. close": json.valors[0].data.Close[i],  //valors.date.Close
            "5. adjusted close": json.valors[0].data.Close[i], //valors.date.Close
            "6. volume": json.valors[0].data.TotalVolume[i] //valors.date.TotalVolume
          }
        }
      }
      object["Monthly Adjusted Time Series"] = value
      return object
    }

    const getDataFromAPI = (symbol) => {
      var url = ""
      if(symbol.exchange === "NYSE"){
        url=`https://www.alphavantage.co/query?function=TIME_SERIES_MONTHLY_ADJUSTED&symbol=${symbol.symbol}&apikey=${process.env.REACT_APP_ALPHAVANTAGE_API_KEY}`
      }else{
        url=`http://localhost:4000/isin/${symbol.symbol}/${symbol.currency}`
      }
        return fetch(url, {mode: 'cors'})
        .then((response) => {
          if (!response.ok) {
            console.log('Looks like there was a problem. Status Code: ' + response.status);
            return;
          } else {
            return response.json()
          }
        })
        .then(json => {
          var dbPromise =  idb.open('stock-price-db', 1, (upgradeDB) => {})
          if(symbol.exchange==="SWX"){
            //Data Structure of data from SWX is diffrent from the one from
            //Alphavantage (NYSE etc.) - this is the reason for this function
            json = convertSWXData(json)
          }
          dbPromise.then( db => {
            var tx = db.transaction('keyval','readwrite')
            var keyValStore = tx.objectStore('keyval')
              //Here I'll add a value to my key-value-store
              //Notice that .put() takes first the value and then the key
              keyValStore.put(json, json["Meta Data"]["2. Symbol"])
          })
          return json["Monthly Adjusted Time Series"]
        })
        .catch(e => {console.log(e)})
      }

    return new Promise (
      (resolve, reject) => {
        var allData = { }
        for(let symbol of symbols){
          allData[symbol.symbol] = idb.open('stock-price-db', 1, (upgradeDB) => {
            //This function is only performed, if there is no indexedDB with the name 'stock-price-db'
            //Here I'll create a key-value-store
            upgradeDB.createObjectStore('keyval')
            //keyValStore.put(undefined, 'shareprices')
          })
          .then((db) => {
            //get Data from IndexedDB
            var tx = db.transaction('keyval')
            var keyValStore = tx.objectStore('keyval')
            return keyValStore.get(symbol.symbol)
          })
          .then((data) => {
            //Check if DB contains data i need
            if(data){
              //returns data
              return data["Monthly Adjusted Time Series"]
            }else{
              //returns promise to deliver data
              return getDataFromAPI(symbol)
            }
          })
          .catch(e => console.log(e))
        }
        resolve(allData)
      }
    )
  }

  render() {
    return (
        <LineChart loadingStatus={this.state.shareprice} data={this.state.allData} symbols={this.state.symbols}/>
    )
  }
}

const mapStateToProps = (state, props) => ({
  ...state
});

export default connect(mapStateToProps)(LineChartWithData)
