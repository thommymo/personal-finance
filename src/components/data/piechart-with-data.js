import React, { Component } from 'react'
import { portfolio, currency, holdingsWithMarketPrice, shareValue } from '../../data/data'
import PieChart from '../../components/molecules/piechart'
import { connect } from 'react-redux'
import TableWithHoldings from '../molecules/tablewithholdings'
import TableWithHoldingsFinMarkets from '../molecules/tablewithholdingsfinmarkets'
import { setPortfolioSelection } from '../../components/data/actions'
import LineChartWithData from '../../components/data/linechartwithdata'

//Data will come from a database later, instead of the import from "../../data/data"

/*
TODO
//Check which steps are really necessary
2. If no Data is available: Show error message
3. If no Data in IndexedDB, but data is loading: Show Loading message
4. If Data in IndexedDB, show with old data and load new data in background
5. If New Data from API, show data
*/

class PieChartWithData extends Component {
  setPortfolioSelection(holdingsType,color){
    this.props.dispatch(setPortfolioSelection(holdingsType,color))
  }
  render() {
    console.log(this.props);
    const { holdingsType, color } = this.props.selection
    const exchangeRates = this.props.exchangeRates.rates
    return (
      <div>
        { exchangeRates.isFetching &&
          //TODO: Add a beautiful Loading state
          <div>Loading</div>
        }
        { !exchangeRates.isFetching && !exchangeRates &&
          //TODO: Add a beautiful Error state
          <div>Error</div>
        }
        { !exchangeRates.isFetching && exchangeRates &&
          <PieChart
            setPortfolioSelection={(holdingsType,color)=>this.setPortfolioSelection(holdingsType,color)}
            portfolio={portfolio}
            currency={exchangeRates}
            holdingsWithMarketPrice={holdingsWithMarketPrice}
            shareValue={shareValue}
          />
        }
        { !exchangeRates.isFetching && exchangeRates && holdingsType && !holdingsWithMarketPrice.some(holding => holding===holdingsType) &&
          <TableWithHoldings
            portfolioSelection={holdingsType}
            color={color}
            portfolio={portfolio}
            currency={exchangeRates}
            holdingsWithMarketPrice={holdingsWithMarketPrice}
            shareValue={shareValue}
          />
        }
        { !exchangeRates.isFetching && exchangeRates && holdingsType && holdingsWithMarketPrice.some(holding => holding===holdingsType) &&
          <div>
            <TableWithHoldingsFinMarkets
              portfolioSelection={holdingsType}
              color={color}
              shareValue={shareValue}
              portfolio={portfolio}
              exchangeRates={exchangeRates}
              currency={exchangeRates}
              holdingsWithMarketPrice={holdingsWithMarketPrice}
              shareValue={shareValue}
            />
            <LineChartWithData symbols={portfolio.filter(holding => holding.type===this.props.selection.holdingsType)}/>
          </div>

        }
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  ...state
});

export default connect(mapStateToProps)(PieChartWithData)
