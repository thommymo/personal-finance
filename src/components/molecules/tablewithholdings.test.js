import React from 'react';
import TableWithHoldings from './tablewithholdings';
import renderer from 'react-test-renderer'
import { portfolio, currency, shareValue } from '../../data/data'

it('Shows portfolio as pie chart', () => {
  const component = renderer.create(
    <TableWithHoldings
      portfolioSelection="Cash"
      color="blue"
      portfolio={portfolio}
      currency={currency}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})
