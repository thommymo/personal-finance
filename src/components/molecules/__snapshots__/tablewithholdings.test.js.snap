// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shows portfolio as pie chart 1`] = `
<table
  className="sc-bdVaJa fBygAm"
  color="blue"
>
  <thead>
    <tr
      className="sc-bwzfXH blIGVZ"
    >
      <th
        className="sc-htpNat eTmyiO"
      >
        Name
      </th>
      <th
        className="sc-htpNat eTmyiO"
      >
        Value in CHF
      </th>
      <th
        className="sc-htpNat eTmyiO"
      >
        Interest Rate
      </th>
      <th
        className="sc-htpNat eTmyiO"
      >
        2018
      </th>
      <th
        className="sc-htpNat eTmyiO"
      >
        After Inflation
      </th>
      <th
        className="sc-htpNat eTmyiO"
      />
    </tr>
  </thead>
  <tbody>
    <tr
      className="sc-bwzfXH blIGVZ"
    >
      <td
        className="sc-bxivhb jSiPdZ"
      >
        DKB EUR Privatkonto
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF986.07
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        0.20%
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF1.97
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        -CHF4.76
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Remove
        </button>
      </td>
    </tr>
    <tr
      className="sc-bwzfXH blIGVZ"
    >
      <td
        className="sc-bxivhb jSiPdZ"
      >
        Postfinance Privatkonto 2
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF3,361.00
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        0.01%
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF0.34
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        -CHF26.55
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Remove
        </button>
      </td>
    </tr>
    <tr
      className="sc-bwzfXH blIGVZ"
    >
      <td
        className="sc-bxivhb jSiPdZ"
      >
        Postfinance Privatkonto
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF13,402.00
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        0.01%
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF1.34
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        -CHF105.88
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Remove
        </button>
      </td>
    </tr>
    <tr
      className="sc-bwzfXH blIGVZ"
    >
      <td
        className="sc-bxivhb jSiPdZ"
      >
        Postfinance Sparkonto
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF60,806.00
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        0.01%
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF6.08
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        -CHF480.37
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Remove
        </button>
      </td>
    </tr>
    <tr
      className="sc-bwzfXH blIGVZ"
    >
      <td
        className="sc-bxivhb jSiPdZ"
      >
        True Wealth GBP Account
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF54.21
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        0.00%
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF0.00
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        -CHF0.33
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Remove
        </button>
      </td>
    </tr>
    <tr
      className="sc-bwzfXH blIGVZ"
    >
      <td
        className="sc-bxivhb jSiPdZ"
      >
        True Wealth USD Account
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF391.08
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        0.00%
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF0.00
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        -CHF3.20
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Remove
        </button>
      </td>
    </tr>
    <tr
      className="sc-bwzfXH blIGVZ"
    >
      <td
        className="sc-bxivhb jSiPdZ"
      >
        True Wealth EUR Account
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF100.83
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        0.00%
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF0.00
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        -CHF0.69
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Remove
        </button>
      </td>
    </tr>
    <tr
      className="sc-bwzfXH blIGVZ"
    >
      <td
        className="sc-bxivhb jSiPdZ"
      >
        True Wealth CHF Account
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF41.00
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        0.00%
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF0.00
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        -CHF0.33
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Remove
        </button>
      </td>
    </tr>
    <tr
      className="sc-bwzfXH blIGVZ"
    >
      <td
        className="sc-bxivhb jSiPdZ"
      >
        ABL kurzfristig
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF20,300.00
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        0.10%
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        CHF20.30
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        -CHF142.10
      </td>
      <td
        className="sc-ifAKCX fTHyJg"
      >
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="sc-gzVnrw biOjMH"
          onClick={[Function]}
        >
          Remove
        </button>
      </td>
    </tr>
  </tbody>
  <tfoot>
    <tr
      className="sc-bwzfXH blIGVZ"
    >
      <td
        className="sc-EHOje bTmYja"
      >
        Sum
      </td>
      <td
        className="sc-bZQynM kQkXwr"
      >
        CHF99,278.00
      </td>
      <td
        className="sc-bZQynM kQkXwr"
      >
        0.03%
      </td>
      <td
        className="sc-bZQynM kQkXwr"
      >
        CHF30.03
      </td>
      <td
        className="sc-bZQynM kQkXwr"
      >
        -CHF764.19
      </td>
      <td
        className="sc-bZQynM kQkXwr"
      />
    </tr>
  </tfoot>
</table>
`;
